proposals/0000-instance-scoping.rst

```diff
@@ -0,0 +1,190 @@
+Simplify type synonym and type instance declarations
+=======================================================
+
+.. author:: Simon Peyton Jones
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight::
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/386>`_.
+.. sectnum::
+.. contents::
+
+Summary
+---------
+
+In a couple of places GHC is being too clever when dealing with type synonym
+or type-family instance declarations.  This proposal suggests two
+related simplifications, that make the language easier to read, less ad hoc,
+and simpler to implement.
+
+It depends on proposal #326 (Inivisible parameters for declarations).
+This proposal has a draft implementation in MR !3145.
+
+Proposed Change Specification
+-----------------------------
+
+* Change #1.  Introduce the following rule: in
+
+  * a type synonym declaration, or
+  * a type family instance declaration
+  every type variable mentioned on the RHS must be bound on the LHS.
+
+  To match this change, in the user manual, remove the text in
+  `Implicit quantification in type synonyms and type family instances (6.4.11.12)
+  <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/poly_kinds.html>`_,
+  from the beginning of that section down to
+  "Kind variables can also be quantified in visible positions...".
+
+  (Implementation note: see ``Note [Implicit quantification in type synonyms]`` in ``GHC.Rename.HsType``.)
+
+* Change #2.   Introduce the following rule: in a type family instance declaration,
+  the instantiation of the family instance fully determined by the left hand side, without
+  looking at the right hand side.
+
+Examples and motivation
+-----------------------
+
+Change #1
+~~~~~~~~~
+
+Consider::
+
+  type T1 = 'Nothing :: Maybe a
+  type T2 = 'Just ('Nothing :: Maybe a)
+  
+``T1`` is currently legal, yielding ``T :: forall a. Maybe a``. The general rule is that the
+free variables of a *top-level* kind signature on the RHS are brought into scope
+implicitly, and will be quantified in the final kind of the type constructor.
+
+In constrast, ``T2`` is currently illegal, because the kind signature is not at the top level.
+
+With this proposal, both declarations woudl be illegal.  Instead you must write:::
+
+  type T1 @a = 'Nothing :: Maybe a
+  type T2 @a = 'Just ('Nothing :: Maybe a)
+
+so that all the variables occurring on the RHS are bound on the LHS.
+This proposal exploits #326 to allow a nice, simple, uniform rule.
+As the manual says, "The reason for this exception [the strange, ad-hoc,
+top-level-kind-signature rule] is that there may be no other way to bind k".
+
+The same rule applies to type family instances::
+
+   type family F a :: k
+
+   type instance F Int = Any :: k -> k
```

Comment:
To be clear, the `k` in `k -> k` is different from the `k` in `type family F a :: k`, right? Perhaps it'd be a good idea to rename one of these `k`s to make that more obvious

Reply:
done